src/Microsoft.AspNet.Mvc.Core/ModelBinding/JQueryFormValueProviderFactory.cs

```diff
@@ -0,0 +1,114 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Globalization;
+using System.Text;
+using System.Threading.Tasks;
+using Microsoft.AspNet.Http;
+using Microsoft.AspNet.Mvc.Core;
+using Microsoft.Framework.Internal;
+
+namespace Microsoft.AspNet.Mvc.ModelBinding
+{
+    public class JQueryFormValueProviderFactory : IValueProviderFactory
+    {
+        public IValueProvider GetValueProvider([NotNull] ValueProviderFactoryContext context)
+        {
+            var request = context.HttpContext.Request;
+
+            if (request.HasFormContentType)
+            {
+                return new JQueryFormValueProvider(
+                    BindingSource.Form,
+                    () => GetValueCollectionAsync(request),
+                    CultureInfo.CurrentCulture);
+            }
+
+            return null;
+        }
+
+        private static async Task<IDictionary<string, string[]>> GetValueCollectionAsync(HttpRequest request)
+        {
+            var formCollection = await request.ReadFormAsync();
+
+            var dictionary = new Dictionary<string, string[]>(StringComparer.OrdinalIgnoreCase);
+            foreach (var entry in formCollection)
+            {
+                var key = NormalizeJQueryToMvc(entry.Key);
+                dictionary[key] = entry.Value;
+            }
+
+            return dictionary;
+        }
+
+        // This is a helper method for Model Binding over a JQuery syntax.
+        // Normalize from JQuery to MVC keys. The model binding infrastructure uses MVC keys.
+        // x[] --> x
+        // [] --> ""
+        // x[12] --> x[12]
+        // x[field]  --> x.field, where field is not a number
+        private static string NormalizeJQueryToMvc(string key)
+        {
+            if (string.IsNullOrEmpty(key))
+            {
+                return string.Empty;
+            }
+
+            StringBuilder builder = null;
+            var position = 0;
+            while (position < key.Length)
+            {
+                var indexOpen = key.IndexOf('[', position);
+                if (indexOpen == -1)
+                {
+                    if (position == 0)
+                    {
+                        // Fast path, no normalization needed.
+                        // This skips string conversion and allocating the string builder.
+                        return key;
+                    }
```

Comment:
This would be much much more straightforward to pull out and just put ahead of the while. Then you don't need to double-check the `builder`